python-package/xgboost/spark/data.py

```diff
@@ -1,192 +1,166 @@
+
+def create_dmatrix_from_partitions(
+    iterator: Iterator[pd.DataFrame], has_validation: bool
+) -> Tuple[DMatrix, Optional[DMatrix]]:
+    train_data: Dict[str, List[np.ndarray]] = defaultdict(list)
+    valid_data: Dict[str, List[np.ndarray]] = defaultdict(list)
+    n_features: List[int] = [0]
+
+    def append(part: pd.DataFrame, name: str, is_valid: bool) -> None:
+        if name in part.columns:
+            array = part[name]
+            if name == alias.data:
+                array = stack_df(array)
+                if n_features[0] == 0:
+                    n_features[0] = array.shape[1]
+                assert n_features[0] == array.shape[1]
+
+            if is_valid:
+                valid_data[name].append(array)
+            else:
+                train_data[name].append(array)
+
+    def make_blob(part: pd.DataFrame, is_valid: bool) -> None:
+        append(part, alias.data, is_valid)
+        append(part, alias.label, is_valid)
+        append(part, alias.weight, is_valid)
+        append(part, alias.margin, is_valid)
+
+    for part in iterator:
         if has_validation:
-            values_val_list.append(pdf_val["values"].to_list())
-            if train:
-                label_val_list.append(pdf_val["label"].to_numpy())
-            if has_weight:
-                weight_val_list.append(pdf_val["weight"].to_numpy())
-            if has_fit_base_margin or has_predict_base_margin:
-                base_margin_val_list.append(pdf_val["baseMargin"].to_numpy())
-
-    # Construct feature_matrix
-    if expected_feature_dims is None:
-        return [], [], [], []
-
-    # Construct feature_matrix, y and w
-    feature_matrix = np.concatenate(values_list)
-    y = np.concatenate(label_list) if train else None
-    w = np.concatenate(weight_list) if has_weight else None
-    b_m = (
-        np.concatenate(base_margin_list)
-        if (has_fit_base_margin or has_predict_base_margin)
-        else None
-    )
-    if has_validation:
-        feature_matrix_val = np.concatenate(values_val_list)
-        y_val = np.concatenate(label_val_list) if train else None
-        w_val = np.concatenate(weight_val_list) if has_weight else None
-        b_m_val = (
-            np.concatenate(base_margin_val_list)
-            if (has_fit_base_margin or has_predict_base_margin)
-            else None
+            train = part.loc[~part[alias.valid], :]
+            valid = part.loc[part[alias.valid], :]
+        else:
+            train, valid = part, None
+
+        make_blob(train, False)
+        if valid is not None:
+            make_blob(valid, True)
+
+    def make(values: Dict[str, List[np.ndarray]]) -> DMatrix:
+        data = concat_or_none(train_data[alias.data])
+        label = concat_or_none(train_data[alias.label])
+        weight = concat_or_none(train_data[alias.weight])
+        margin = concat_or_none(train_data[alias.margin])
+        return DMatrix(data=data, label=label, weight=weight, base_margin=margin)
```

Comment:
I think you forget to apply other dmatrix params:
https://github.com/dmlc/xgboost/blob/e28f6f66577eb31451e5d3fbf4be0a21626f7372/python-package/xgboost/spark/core.py#L565